### PR TITLE
feat: add sendgrid key detection

### PR DESCRIPTION
**Pull request by mtsalenc**
This rule allows detecting more generic api keys like the ones used by SendGrid.

I cannot run the tests locally or compile the project to continue development due to a typescript error which seems to be present for all packages:
```
TSError: ⨯ Unable to compile TypeScript:
test/index.test.ts:1:26 - error TS2307: Cannot find module '@secretlint/tester'.

1 import { snapshot } from "@secretlint/tester";
                           ~~~~~~~~~~~~~~~~~~~~
test/index.test.ts:18:17 - error TS7006: Parameter 'name' implicitly has an 'any' type.

18     }).forEach((name, test) => {
                   ~~~~
test/index.test.ts:18:23 - error TS7006: Parameter 'test' implicitly has an 'any' type.

18     }).forEach((name, test) => {

...
```

Do you have any tips on how I can fix this?

### CODE REVIEW

**Comment by azu** [comment]
> I cannot run the tests locally or compile the project to continue development due to a typescript error which seems to be present for all packages:

Fmm. 
It is related with #119 changes(not yet released) 

Can you run `yarn install && yarn run build` on root directory?
#119 changes all packages and need to rebuild all packages.
It will build new `@secretlint/types` pacakge in your local.

It is implicitly dependencies.
[Project References](https://www.typescriptlang.org/docs/handbook/project-references.html) will resolve implicitly dependencies and I've created an issue https://github.com/secretlint/secretlint/issues/129

> generic key detection

Would we change this rule to more specific rule like `@secretlint/secret-rule-sendgrid`?
Or, We need to add more strict condition for avoiding false-positive.

We want to reduce false-positive. It is one of [Secretlint's Philosophy](https://github.com/secretlint/secretlint#philosophy).

I think that current implementation has a false-positive.

For example enctrypted key like [SealedSecrets](https://github.com/bitnami-labs/sealed-secrets#sealedsecrets-as-templates-for-secrets)

```
api-key="<encrypted-apikey>"
```

I found similar problem in git-secrets https://github.com/awslabs/git-secrets/blob/5e28df337746db4f070c84f7069d365bfd0d72a8/git-secrets#L239 and I've changed this logic in secretlint-rule-aws https://github.com/secretlint/secretlint/blob/c09ba64dc91dd85cd3066492f1054bcd15457190/packages/%40secretlint/secretlint-rule-aws/src/index.ts#L102-L106

While, Detecting `SG.<ID>.<VALUE>` as Send Grid API Key is reasonable.

- [API Keys - SendGrid Documentation | SendGrid](https://sendgrid.com/docs/API_Reference/Web_API_v3/API_Keys/index.html)
- [How to get the full SendGrid API Key? - Stack Overflow](https://stackoverflow.com/questions/42030912/how-to-get-the-full-sendgrid-api-key)

It is similar with slack api key.
https://github.com/secretlint/secretlint/blob/c09ba64dc91dd85cd3066492f1054bcd15457190/packages/%40secretlint/secretlint-rule-slack/src/index.ts#L41-L47

**Comment by mtsalenc** [comment]
> Can you run yarn install && yarn run build on root directory?

I did that and it seems to be all good now.

> Would we change this rule to more specific rule like @secretlint/secret-rule-sendgrid?
Or, We need to add more strict condition for avoiding false-positive.
We want to reduce false-positive. It is one of Secretlint's Philosophy.
I think that current implementation has a false-positive.

You are right, I can update this to do @secretlint/secret-rule-sendgrid instead :+1:

File: packages/@secretlint/secretlint-rule-sendgrid/src/index.ts [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    SecretLintSourceCode,
+} from "@secretlint/types";
+import { matchPatterns } from "@textlint/regexp-string-matcher";
+
+require("string.prototype.matchall").shim();
+
+export const messages = {
+    SENDGRID_KEY: {
+        en: (props: { KEY: string }) => `found Sendgrid api key: ${props.KEY}`,
+        ja: (props: { KEY: string }) => `Sendgrid APIキーが見つかりました： ${props.KEY}`, // TODO: Needs review.
```
Comment (azu) [inline packages/@secretlint/secretlint-rule-sendgrid/src/index.ts:14]: Looks like good!
You can just remove todo.

```suggestion
        ja: (props: { KEY: string }) => `Sendgrid APIキーが見つかりました： ${props.KEY}`,
```

File: packages/@secretlint/secretlint-rule-sendgrid/src/index.ts [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    options,
+    context,
+    t,
+}: {
+    source: SecretLintSourceCode;
+    options: Required<Options>;
+    context: SecretLintRuleContext;
+    t: SecretLintRuleMessageTranslate<typeof messages>;
+}) {
+    const SENDGRID_KEY_PATTERN = /SG.\w+.\w+-\w+/g;
```
Comment (azu) [inline packages/@secretlint/secretlint-rule-sendgrid/src/index.ts:37]:
```suggestion
    const SENDGRID_KEY_PATTERN = /SG\.\w+\.\w+-\w+/g;
```

Maybe `.` should be escaped, becase `.` match any charater.

Also, Can we limit the `\w` lengths for clarity?

```js
const SENDGRID_KEY_PATTERN = /SG\.\w{1,128}\.\w{1,128}-\w{1,128}/g
```

**Comment by azu** [comment]
I'll publish this pacakge in next update(today ~ tommorow).
Also, I'll add secretlint-rule-sendgrid to recommended preset #131

Thanks for improving!
I've invited @mtsalenc as collaboration into secretlint organization 🎉

**Pull request merged**